### PR TITLE
BUG: groupby.agg returns incorrect results for uint64 cols (#26310)

### LINKED ISSUE #26310: groupby.agg (first, last, min, etc...) returns incorrect results for uint64 columns
#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
import numpy as np

df = pd.DataFrame({'x': 6903052872240755750, 'y': [1,2]})
print(df.groupby('y').agg({'x': 'first'}))
df.x = df.x.astype(np.uint64)
print(df.groupby('y').agg({'x': 'first'}))
```
#### Problem description

groupby.agg (first, last, min, etc...) returns incorrect results for uint64 columns

#### Expected Output

expect same results in both cases below:
                     x
y                     
1  6903052872240755750
2  6903052872240755750
                     x
y                     
1  6903052872240755712
2  6903052872240755712

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.8.final.0
python-bits: 64
OS: Darwin
OS-release: 18.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: None
pip: 19.1
setuptools: 41.0.1
Cython: None
numpy: 1.16.3
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: 3.5.1
numexpr: 2.6.9
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by mahepe**
- [x] closes #26310
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

A way to avoid incorrect coercion to float64 in the case of uint64 input is to introduce an additional check to `ensure_int64_or_float64`

### CODE REVIEW

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -258,6 +258,7 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
 
+- Bug where groupby.agg (first, last, min, etc...) returns incorrect results for uint64 columns. (:issue:`26310`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: move to grouping use full references (see other issues on how to do this)

File: pandas/core/dtypes/common.py [hunk 2]
```diff
@@ -90,9 +90,11 @@ def ensure_categorical(arr):
 def ensure_int64_or_float64(arr, copy=False):
     """
     Ensure that an dtype array of some integer dtype
-    has an int64 dtype if possible
+    has an int64 dtype if possible.
```
Comment (jreback) [inline pandas/core/dtypes/common.py]: can you move these comments in to a Notes section.

File: pandas/core/dtypes/common.py [hunk 3]
```diff
@@ -107,11 +109,16 @@ def ensure_int64_or_float64(arr, copy=False):
...
               possible without overflow.
               Otherwise the input array cast to float64.
+              If the array is explicitly of type uint64 the type
+              will remain unchanged.
     """
     try:
         return arr.astype('int64', copy=copy, casting='safe')
     except TypeError:
-        return arr.astype('float64', copy=copy)
+        try:
```
Comment (jreback) [inline pandas/core/dtypes/common.py]: use a pass rather than a nested try/except

File: pandas/core/dtypes/common.py [hunk 4]
```diff
@@ -107,11 +109,16 @@ def ensure_int64_or_float64(arr, copy=False):
     out_arr : The input array cast as int64 if
```
Comment (jreback) [inline pandas/core/dtypes/common.py]: make this simpler; these statements are for the Notes section

File: pandas/core/dtypes/common.py [hunk 5]
```diff
@@ -90,9 +90,11 @@ def ensure_categorical(arr):
 def ensure_int64_or_float64(arr, copy=False):
```
Comment (jreback) [inline pandas/core/dtypes/common.py]: can you type this (ArrayLike) -> np.array

File: pandas/core/dtypes/common.py [hunk 6]
```diff
@@ -90,9 +90,11 @@ def ensure_categorical(arr):
...
     """
     Ensure that an dtype array of some integer dtype
```
Comment (jreback) [inline pandas/core/dtypes/common.py]: rename this to: ensure_int_or_float

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 7]
```diff
@@ -313,3 +313,14 @@ def test_order_aggregate_multiple_funcs():
     expected = pd.Index(['sum', 'max', 'mean', 'ohlc', 'min'])
 
     tm.assert_index_equal(result, expected)
+
+
+def test_uint64_type_handling():
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: can you add a battery of tests in pandas/tests/dtypes/test_common.py I don't think we have this tested at all.

Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: can you parameterize on first, last. Does this also deal with min, max? if so pls add parameterizeation as well. Anything else?

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 8]
```diff
@@ -313,3 +313,14 @@ def test_order_aggregate_multiple_funcs():
...
+    # GH 26310
+    df1 = pd.DataFrame({'x': 6903052872240755750, 'y': [1, 2]})
+    df1.groupby('y').agg({'x': 'first'})
+    df2 = df1
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: use

result =
expected = 
tm.assert_frame_equal(result, expected)

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 9]
```diff
@@ -313,3 +313,14 @@ def test_order_aggregate_multiple_funcs():
...
+    df2.x = df2.x.astype(np.uint64)
+    df2.groupby('y').agg({'x': 'first'})
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: we are going to want to parameterize this on an int64 as well (so another later of paramaterization)

**Review by jreback — CHANGES_REQUESTED** [review]
thanks. this needs a bit of testing; my comments should be straightforward to handle.

File: doc/source/whatsnew/v0.25.0.rst (hunk 1 above)
Comment (mahepe) [reply doc/source/whatsnew/v0.25.0.rst]: done

File: pandas/core/dtypes/common.py (hunk 2 above)
Comment (mahepe) [reply pandas/core/dtypes/common.py]: done

File: pandas/core/dtypes/common.py (hunk 3 above)
Comment (mahepe) [reply pandas/core/dtypes/common.py]: done

File: pandas/core/dtypes/common.py (hunk 4 above)
Comment (mahepe) [reply pandas/core/dtypes/common.py]: done (moved what I added to the notes section)

File: pandas/core/dtypes/common.py (hunk 5 above)
Comment (mahepe) [reply pandas/core/dtypes/common.py]: done

File: pandas/core/dtypes/common.py (hunk 6 above)
Comment (mahepe) [reply pandas/core/dtypes/common.py]: done

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 7 above)
Comment (mahepe) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: I did add the parameterization as asked below. I wasn't sure if you meant I should add new tests in addition to `test_uint64_type_handling`

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 8 above)
Comment (mahepe) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: done

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 9 above)
Comment (mahepe) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: done

File: pandas/core/dtypes/common.py [hunk 10]
```diff
@@ -87,10 +88,10 @@ def ensure_categorical(arr):
     return arr
 
 
-def ensure_int64_or_float64(arr, copy=False):
+def ensure_int_or_float(arr: ArrayLike, copy=False) -> np.array:
```
Comment (WillAyd) [inline pandas/core/dtypes/common.py]: Nice annotation!

File: doc/source/whatsnew/v0.25.0.rst [hunk 11]
```diff
@@ -403,6 +403,7 @@ Groupby/Resample/Rolling
 - Bug in :meth:`pandas.core.groupby.GroupBy.idxmax` and :meth:`pandas.core.groupby.GroupBy.idxmin` with datetime column would return incorrect dtype (:issue:`25444`, :issue:`15306`)
 - Bug in :meth:`pandas.core.groupby.GroupBy.cumsum`, :meth:`pandas.core.groupby.GroupBy.cumprod`, :meth:`pandas.core.groupby.GroupBy.cummin` and :meth:`pandas.core.groupby.GroupBy.cummax` with categorical column having absent categories, would return incorrect result or segfault (:issue:`16771`)
 - Bug in :meth:`pandas.core.groupby.GroupBy.nth` where NA values in the grouping would return incorrect results (:issue:`26011`)
+- Bug in :meth:`pandas.core.groupby.BaseGrouper._cython_operation` where incorrect results are returned for uint64 columns. (:issue:`26310`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.25.0.rst]: Whatsnew messages should be user-facing, i.e. only references items exposed via the API. This isn't, but I think you can refactor to reference `pandas.core.groupby.GroupBy.agg` instead

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 12]
```diff
@@ -313,3 +313,16 @@ def test_order_aggregate_multiple_funcs():
...
+@pytest.mark.parametrize('dtype', [np.int64, np.uint64])
+@pytest.mark.parametrize('how', ['first', 'last', 'min',
+                                 'max', 'mean', 'median'])
+def test_uint64_type_handling(dtype, how):
+    # GH 26310
+    df = pd.DataFrame({'x': 6903052872240755750, 'y': [1, 2]})
+    expected = df.groupby('y').agg({'x': how})
+    df.x = df.x.astype(dtype)
+    result = df.groupby('y').agg({'x': how})
+    result.x = result.x.astype(np.int64)
```
Comment (WillAyd) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: Can we add test coverage for a value that exceeds the upper limit of an int64?

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 13]
```diff
@@ -313,3 +313,16 @@ def test_order_aggregate_multiple_funcs():
     expected = pd.Index(['sum', 'max', 'mean', 'ohlc', 'min'])
 
     tm.assert_index_equal(result, expected)
+
+
+@pytest.mark.parametrize('dtype', [np.int64, np.uint64])
+@pytest.mark.parametrize('how', ['first', 'last', 'min',
```
Comment (WillAyd) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: Just out of curiosity how did you select these methods? The issue would cover more than just these right?

Just asking as at some point we should probably add a shared fixture for the different aggregation / transformation methods. Probably a separate PR but just curious if this subset is intentional

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/common.py (hunk 10 above)
Comment (mahepe) [reply pandas/core/dtypes/common.py]: hah thanks, it was actually due to the other reviewer :)

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 13 above)
Comment (mahepe) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: I tried to select a set that would satisfy the requests by the other reviewer. Since this PR alters a pretty generic method, I'm sure you could test for a way larger set of methods. It's just that I don't know the codebase well enough to design such tests.

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 12 above)
Comment (mahepe) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: It's not immediately clear to me what we would be testing for in that case. Issue #26310 is basically that there is some method `f` such that `f(x) != f(y)`, even though `x == y`, when `x` is of type `np.int64` and `y` is of type `np.uint64`.

If the value of `y` would exceed the upper limit of `np.int64` you couldn't represent it as an `np.int64` (?) and I'm not sure how you could pick `x` in that case to produce the above situation.

If I have misunderstood, please elaborate.

File: doc/source/whatsnew/v0.25.0.rst (hunk 11 above)
Comment (mahepe) [reply doc/source/whatsnew/v0.25.0.rst]: fixed, thanks!

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 12 above)
Comment (WillAyd) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: The problem here is going to be with unsigned integers in the range of 2**63 through 2**64-1 (range where uint64 exceeds int64). I'm not sure if that would even work with these agg functions and don't want to open up a Pandora's box here but if that doesn't work we might just need to reword the whatsnew

Comment (mahepe) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: I suggest rewording whatsnew here and dealing with that in another pull request. What do you think @WillAyd? How would you like the whatsnew to be worded?

File: pandas/core/dtypes/common.py [hunk 14]
```diff
@@ -107,9 +108,18 @@ def ensure_int64_or_float64(arr, copy=False):
     out_arr : The input array cast as int64 if
               possible without overflow.
               Otherwise the input array cast to float64.
+
+    Notes
+    -------
```
Comment (jreback) [inline pandas/core/dtypes/common.py]: the underline should be the same length as Notes

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 12 above)
Comment (jreback) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: hmm, input uint64's *should* work with your code change, but @mahepe ok to make an issue about this.

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. pls merge master; there is a conflict in the whatsnew. some doc checks are failing as well. ping on green.

File: pandas/core/dtypes/common.py (hunk 14 above)
Comment (mahepe) [reply pandas/core/dtypes/common.py]: done

**Comment by mahepe** [comment]
all green @jreback !

**Review by WillAyd — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mahepe
